From 7ed4574eb66cda0dcbecf3185399b571c3b01693 Mon Sep 17 00:00:00 2001
From: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
Date: Sat, 16 Dec 2023 16:01:54 +0900
Subject: [PATCH] Avoid detecting OpenSSL 3.2 as quictls

Previously, SSL_is_quic is used to detect quictls.  But OpenSSL 3.2
has added it as a part of its own QUIC implementation which breaks our
script.  Use SSL_provide_quic_data to correctly identify quictls and
avoid OpenSSL 3.2.
---
 CMakeLists.txt | 6 +++---
 configure.ac   | 5 ++---
 2 files changed, 5 insertions(+), 6 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 085a59ee..a1949e5f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -151,13 +151,13 @@ if(ENABLE_OPENSSL AND OPENSSL_FOUND)
   if(WIN32)
     set(CMAKE_REQUIRED_LIBRARIES "${CMAKE_REQUIRED_LIBRARIES}" "ws2_32" "bcrypt")
   endif()
-  check_symbol_exists(SSL_is_quic "openssl/ssl.h" HAVE_SSL_IS_QUIC)
-  if(NOT HAVE_SSL_IS_QUIC)
+  check_symbol_exists(SSL_provide_quic_data "openssl/ssl.h" HAVE_SSL_PROVIDE_QUIC_DATA)
+  if(NOT HAVE_SSL_PROVIDE_QUIC_DATA)
     message(WARNING "Disabling OpenSSL due to lack of QUIC support in ${OPENSSL_LIBRARIES}")
   endif()
   cmake_pop_check_state()
 endif()
-if(ENABLE_OPENSSL AND HAVE_SSL_IS_QUIC)
+if(ENABLE_OPENSSL AND HAVE_SSL_PROVIDE_QUIC_DATA)
   set(OPENSSL_INCLUDE_DIRS  ${OPENSSL_INCLUDE_DIR})
   set(HAVE_OPENSSL TRUE)
   set(HAVE_CRYPTO TRUE)
diff --git a/configure.ac b/configure.ac
index df49cd63..eaaf778d 100644
--- a/configure.ac
+++ b/configure.ac
@@ -252,12 +252,11 @@ if test "x${request_openssl}" != "xno"; then
     CFLAGS="$OPENSSL_CFLAGS $CFLAGS"
     LIBS="$OPENSSL_LIBS $LIBS"
 
-    AC_MSG_CHECKING([for SSL_is_quic])
+    AC_MSG_CHECKING([for SSL_provide_quic_data])
     AC_LINK_IFELSE([AC_LANG_PROGRAM([[
       #include <openssl/ssl.h>
     ]], [[
-      SSL *ssl = NULL;
-      SSL_is_quic(ssl);
+      SSL_provide_quic_data(NULL, 0, NULL, 0);
     ]])],
     [AC_MSG_RESULT([yes]); have_openssl_quic=yes],
     [AC_MSG_RESULT([no]); have_openssl_quic=no])
